rust-symcrypt/src/ecdh.rs

```diff
@@ -0,0 +1,269 @@
+/* EcDh functions. For further documentation please refer to symcrypt.h */
+
+use crate::curve_type::*;
+use crate::errors::SymCryptError;
+use std::vec;
+use symcrypt_sys;
+
+pub struct EcDh {
+    curve_type: CurveType,
+    expanded_curve: Option<symcrypt_sys::PSYMCRYPT_ECURVE>,
+    key: Option<symcrypt_sys::PSYMCRYPT_ECKEY>,
+}
+
+impl EcDh {
+    /// EcDh::new() returns a EcDh struct that has a private/public key pair.
+    pub fn new(curve: CurveType) -> Result<Self, SymCryptError> {
+        let mut instance = EcDh {
+            curve_type: curve,
+            expanded_curve: None,
+            key: None,
+        };
+
+        unsafe {
+            // SAFETY: FFI calls
+            let curve_ptr =
+                symcrypt_sys::SymCryptEcurveAllocate(convert_curve(instance.curve_type), 0);
+            if curve_ptr.is_null() {
+                return Err(SymCryptError::MemoryAllocationFailure);
+            }
+
+            instance.expanded_curve = Some(curve_ptr);
+
+            let key_ptr = symcrypt_sys::SymCryptEckeyAllocate(curve_ptr);
+            if key_ptr.is_null() {
+                return Err(SymCryptError::MemoryAllocationFailure);
+            }
+
+            instance.key = Some(key_ptr);
+
+            match symcrypt_sys::SymCryptEckeySetRandom(
+                symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDH,
+                instance.key.unwrap(),
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => {
+                    Ok(instance)
+                }
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    /// EcDh::from_public_key_bytes()    function returns a EcDh struct that only has a public key attached.
+    pub fn from_public_key_bytes(
+        curve: CurveType,
+        public_key: &[u8],
+    ) -> Result<Self, SymCryptError> {
+        let mut instance = EcDh {
+            curve_type: curve,
+            expanded_curve: None,
+            key: None,
+        };
+
+        let num_format = get_num_format(instance.curve_type);
+
+        let ec_point_format = symcrypt_sys::_SYMCRYPT_ECPOINT_FORMAT_SYMCRYPT_ECPOINT_FORMAT_XY;
+
+        unsafe {
+            // SAFETY: FFI calls
+            let curve_ptr =
+                symcrypt_sys::SymCryptEcurveAllocate(convert_curve(instance.curve_type), 0);
+            if curve_ptr.is_null() {
+                return Err(SymCryptError::MemoryAllocationFailure);
+            }
+
+            instance.expanded_curve = Some(curve_ptr);
```

Comment:
Can `expanded_curve` every be `None` from this point on? If not, I'd recommend structuring this function sort of like:

```rust 
let curve_type = curve;
let expanded_curve = SymCryptEcurveAllocate(...);
let mut key_ptr = ptr::null();
let key = match SymCryptEckeySetValue(...) {
    NO_ERROR => key_ptr,
    err => return Err(err.into()),
}

EcDh {
    curve_type,
    expanded_curve,
    key
}
```

This way you don't have to worry about unwrapping the fields later on since they are guaranteed always to be valid.

Reply:
In the above err => we would also need to call 
symcrypt_sys::SymCryptEcurveFree(expanded_curve)

Reply:
Also, we were using the instance data structure to hold the pointers that would get freed for any errors in the SymCryptEckeySetValue call.